Corrupted snapshot when adding a new instance?

[This code](https://github.com/jerrinot/jetexper/blob/master/src/main/java/info/jerrinot/jetexper/Main.java) reliably fails with:
```
Exception in thread "main" java.util.concurrent.CompletionException: com.hazelcast.jet.JetException: State for job 'myJob', execution d9fa-ce2c-9188-f677 in '__jet.snapshot.c03f-1517-ca24-b017.1' corrupted: it should have 3 entries, but has 10 entries
	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
	at com.hazelcast.jet.Job.join(Job.java:104)
	at info.jerrinot.jetexper.Main.main(Main.java:78)
Caused by: com.hazelcast.jet.JetException: State for job 'myJob', execution d9fa-ce2c-9188-f677 in '__jet.snapshot.c03f-1517-ca24-b017.1' corrupted: it should have 3 entries, but has 10 entries
	at com.hazelcast.jet.impl.SnapshotValidator.validateSnapshot(SnapshotValidator.java:52)
	at com.hazelcast.jet.impl.MasterContext.rewriteDagWithSnapshotRestore(MasterContext.java:510)
	at com.hazelcast.jet.impl.MasterContext.tryStartJob(MasterContext.java:474)
	at com.hazelcast.jet.impl.JobCoordinationService.tryStartJob(JobCoordinationService.java:711)
	at com.hazelcast.jet.impl.JobCoordinationService.restartJob(JobCoordinationService.java:557)
	at com.hazelcast.jet.impl.MasterContext.lambda$finalizeJob$16(MasterContext.java:937)
	at com.hazelcast.jet.impl.MasterContext.finalizeJob(MasterContext.java:970)
	at com.hazelcast.jet.impl.MasterContext.lambda$onCompleteExecutionCompleted$14(MasterContext.java:901)
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:176)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1603)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:250)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

I am aware sources created via `SourceBuilder` do not provide any processing guarantees, but still I would not expect this.

I am on c268f2dced

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/SnapshotValidationRecord.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/SnapshotValidationRecord.java
@@ -54,7 +54,7 @@ public class SnapshotValidationRecord implements IdentifiedDataSerializable {
         return snapshotId;
     }
 
-    long numChunks() {
+    public long numChunks() {
         return numChunks;
     }
 

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/SnapshotValidator.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/SnapshotValidator.java
@@ -49,7 +49,7 @@ final class SnapshotValidator {
                     && ((SnapshotDataKey) e.getKey()).snapshotId() == snapshotId);
             if (validationRecord.numChunks() != filteredCount) {
                 throw new JetException("State for " + jobIdString + " in '" + map.getName() + "' corrupted: it should " +
-                        "have " + validationRecord.numChunks() + " entries, but has " + (map.size() - 1) + " entries");
+                        "have " + validationRecord.numChunks() + " entries, but has " + (map.size() - 1));
             }
         }
         if (snapshotId != NO_SNAPSHOT && snapshotId != validationRecord.snapshotId()) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java
@@ -146,6 +146,7 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
         if (length > usableChunkSize) {
             return putAsyncToMap(partitionId, () -> {
                 byte[] data = new byte[serializedByteArrayHeader.length + length + valueTerminator.length];
+                totalKeys++;
                 int offset = 0;
                 System.arraycopy(serializedByteArrayHeader, 0, data, offset, serializedByteArrayHeader.length);
                 offset += serializedByteArrayHeader.length - Bits.INT_SIZE_IN_BYTES;
@@ -157,7 +158,9 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
                 offset += entry.getKey().totalSize() - HeapData.TYPE_OFFSET;
 
                 copyWithoutHeader(entry.getValue(), data, offset);
-                System.arraycopy(valueTerminator, 0, data, length, valueTerminator.length);
+                offset += entry.getValue().totalSize() - HeapData.TYPE_OFFSET;
+
+                System.arraycopy(valueTerminator, 0, data, offset, valueTerminator.length);
 
                 return new HeapData(data);
             });
@@ -203,8 +206,6 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
         buffer.write(valueTerminator, 0, valueTerminator.length);
         final byte[] data = buffer.toByteArray();
         updateSerializedBytesLength(data);
-        totalPayloadBytes += buffer.size;
-        totalChunks++;
         buffer.reset();
         buffer.write(serializedByteArrayHeader, 0, serializedByteArrayHeader.length);
         return new HeapData(data);
@@ -227,9 +228,12 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
         }
         try {
             // we put a Data instance to the map directly to avoid the serialization of the byte array
+            Data data = dataSupplier.get();
+            totalPayloadBytes += data.dataSize();
+            totalChunks++;
             ICompletableFuture<Object> future = currentMap.putAsync(
                     new SnapshotDataKey(partitionKeys[partitionId], currentSnapshotId, vertexName, partitionSequence),
-                    dataSupplier.get());
+                    data);
             partitionSequence += memberCount;
             future.andThen(callback);
             numActiveFlushes.incrementAndGet();

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java
@@ -49,7 +49,7 @@ import java.util.function.Supplier;
 
 public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
 
-    private static final int DEFAULT_CHUNK_SIZE = 128 * 1024;
+    public static final int DEFAULT_CHUNK_SIZE = 128 * 1024;
 
     final int usableChunkSize; // this includes the serialization header for byte[], but not the terminator
     final byte[] serializedByteArrayHeader = new byte[3 * Bits.INT_SIZE_IN_BYTES];

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/execution/SnapshotLargeChunk_IntegrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/execution/SnapshotLargeChunk_IntegrationTest.java
@@ -1,0 +1,96 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.execution;
+
+import com.hazelcast.core.IMap;
+import com.hazelcast.jet.JetInstance;
+import com.hazelcast.jet.Job;
+import com.hazelcast.jet.config.JobConfig;
+import com.hazelcast.jet.config.ProcessingGuarantee;
+import com.hazelcast.jet.core.AbstractProcessor;
+import com.hazelcast.jet.core.DAG;
+import com.hazelcast.jet.core.JetTestSupport;
+import com.hazelcast.jet.impl.JobRepository;
+import com.hazelcast.jet.impl.SnapshotValidationRecord;
+import com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl;
+import com.hazelcast.nio.Bits;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import javax.annotation.Nonnull;
+import java.util.IntSummaryStatistics;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.jet.core.JobStatus.RUNNING;
+import static java.util.concurrent.TimeUnit.DAYS;
+import static java.util.stream.Collectors.summarizingInt;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastSerialClassRunner.class)
+public class SnapshotLargeChunk_IntegrationTest extends JetTestSupport {
+
+    @Test
+    public void test_snapshotRestoreLargeChunk() {
+        JetInstance instance = createJetMember();
+        DAG dag = new DAG();
+        dag.newVertex("src", LargeStateP::new).localParallelism(1);
+        Job job = instance.newJob(dag, new JobConfig()
+                .setProcessingGuarantee(ProcessingGuarantee.EXACTLY_ONCE)
+                .setSnapshotIntervalMillis(DAYS.toMillis(1)));
+
+        assertJobStatusEventually(job, RUNNING);
+        job.restart();
+        assertJobStatusEventually(job, RUNNING);
+
+        // assert that the snapshot exists and that the chunk was large enough
+        IMap<Object, Object> map = instance.getMap(JobRepository.snapshotDataMapName(job.getId(), 0));
+        SnapshotValidationRecord validationRec = (SnapshotValidationRecord) map.get(SnapshotValidationRecord.KEY);
+        assertEquals(1, validationRec.numChunks());
+        IntSummaryStatistics stats = map.values().stream()
+                                          .filter(v -> v instanceof byte[])
+                                          .collect(summarizingInt(v -> ((byte[]) v).length));
+        assertTrue("min=" + stats.getMin(), stats.getMin() > AsyncSnapshotWriterImpl.DEFAULT_CHUNK_SIZE);
+    }
+
+    /**
+     * Processor that saves large value to the key and asserts it when restoring.
+     */
+    private static final class LargeStateP extends AbstractProcessor {
+
+        private final int[] data = IntStream.range(0, AsyncSnapshotWriterImpl.DEFAULT_CHUNK_SIZE / Bits.INT_SIZE_IN_BYTES)
+                                            .toArray();
+
+        @Override
+        public boolean complete() {
+            return false;
+        }
+
+        @Override
+        public boolean saveToSnapshot() {
+            return tryEmitToSnapshot("key", data);
+        }
+
+        @Override
+        protected void restoreFromSnapshot(@Nonnull Object key, @Nonnull Object value) {
+            assertEquals("key", key);
+            assertArrayEquals(data, (int[]) value);
+        }
+    }
+}

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImplTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImplTest.java
@@ -30,6 +30,8 @@ import com.hazelcast.jet.impl.JetService;
 import com.hazelcast.jet.impl.execution.SnapshotContext;
 import com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.CustomByteArrayOutputStream;
 import com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.SnapshotDataKey;
+import com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.SnapshotDataValueTerminator;
+import com.hazelcast.nio.Bits;
 import com.hazelcast.nio.BufferObjectDataInput;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.spi.impl.NodeEngineImpl;
@@ -46,6 +48,7 @@ import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import java.io.ByteArrayOutputStream;
+import java.io.IOException;
 import java.io.Serializable;
 import java.util.Map.Entry;
 import java.util.stream.Stream;
@@ -171,14 +174,27 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
     }
 
     @Test
-    public void when_singleLargeEntry_then_flushedImmediately() {
+    public void when_singleLargeEntry_then_flushedImmediatelyAndDeserializesCorrectly() throws IOException {
         // When
-        Entry<Data, Data> entry = entry(serialize("k"), serialize(generate(() -> "a").limit(128).collect(joining())));
+        String key = "k";
+        String value = generate(() -> "a").limit(128).collect(joining());
+        Entry<Data, Data> entry = entry(serialize(key), serialize(value));
         assertTrue("entry not longer than usable chunk size", serializedLength(entry) > writer.usableChunkSize);
         assertTrue(writer.offer(entry));
 
         // Then
-        assertTargetMapEntry("k", 0, serializedLength(entry));
+        assertTargetMapEntry(key, 0, serializedLength(entry));
+        assertEquals(1, writer.getTotalChunks());
+        assertEquals(1, writer.getTotalKeys());
+
+        // Then2 - try to deserialize the entry
+        int partitionKey = writer.partitionKey(partitionService.getPartitionId(key));
+        byte[] data = map.get(new SnapshotDataKey(partitionKey, 1, "vertex", 0));
+        assertEquals(data.length + Bits.INT_SIZE_IN_BYTES, writer.getTotalPayloadBytes());
+        BufferObjectDataInput in = serializationService.createObjectDataInput(data);
+        assertEquals(key, in.readObject());
+        assertEquals(value, in.readObject());
+        assertEquals(SnapshotDataValueTerminator.INSTANCE, in.readObject());
     }
 
     @Test

